Makefile

```diff
@@ -1,130 +1,252 @@
+SHELL := bash
+.ONESHELL:
+.SHELLFLAGS := -o errexit -o nounset -o pipefail -c
+
+.DEFAULT_GOAL := all
+
+# Providers can be built without the SDKs, and SDKs can be built
+# without the providers.
+.PHONY: all
+all: build-all-providers
+all: sdks
+
+# Build Provider
+########################################################################
+
+bin/$(tfgen): provider/cmd/pulumi-tfgen-hcp/main.go
+# TODO: fold this command into the goreleaser file?
+	cd provider
+	go build \
+		-o $(CURDIR)/bin/$(tfgen) \
+		-ldflags "-X $(project)/$(version_path)=$(version)" \
+		$(project)/provider/cmd/$(tfgen)
+
+provider/cmd/$(provider)/schema.json: install-plugins
+provider/cmd/$(provider)/schema.json: bin/$(tfgen)
+provider/cmd/$(provider)/schema.json: provider/resources.go
+provider/cmd/$(provider)/schema.json: provider/shim/go.mod
+	cd provider
+	$(CURDIR)/bin/$(tfgen) schema -v=2 --out cmd/$(provider)
+
+provider/cmd/$(provider)/schema.go: provider/cmd/$(provider)/schema.json
+	cd provider
+	VERSION=$(version) go generate -v -x cmd/$(provider)/main.go
+
+.PHONY: lint
+# We must have schema.go present for our code to lint properly
+lint: provider/cmd/$(provider)/schema.go
+lint: # Lint the provider code
+	(cd provider && $(GOLANGCI_LINT) run -v)
+	(cd examples && $(GOLANGCI_LINT) run -v)
+
+.PHONY: format
+format: # Format all code
+	gofmt -s -w provider
+	gofmt -s -w examples
+
+.PHONY: build-all-providers
+build-all-providers: export GORELEASER_CURRENT_TAG=$(version)
+build-all-providers: ## Just to ensure we can build across all platforms
+	goreleaser build --snapshot --rm-dist
+
+# NOTE: we actually don't *need* this, since nothing else depends on
+# it This is a simplification of the corresponding target from the
+# Pulumi Terraform Provider boilerplate repository, using `goreleaser`
+# to build the provider only for the current operating system. If it
+# truly turns out that we don't ever have a need for this, we can
+# remove it. However, for now, it has the nice effect of consolidating
+# all build logic within the .goreleaser.yml file.
+#
+# TODO: This may be necessary for running tests, though.
+
+bin/$(provider): export GORELEASER_CURRENT_TAG=$(version)
+bin/$(provider):
+	goreleaser build \
+		--output=bin/$(provider) \
+		--single-target \
+		--snapshot \
+		--rm-dist
+
+.PHONY: provider
+provider: bin/$(provider)
+
+# Build all SDKs
+########################################################################
+
+sdk_deps := bin/$(tfgen) provider/cmd/$(provider)/schema.go
+
+.PHONY: sdks
+sdks: $(addsuffix -sdk,$(sdks))
+sdks: # Build all the SDKs
+
+.PHONY: nodejs-sdk
+nodejs-sdk: $(sdk_deps)
+nodejs-sdk: # Build the Node SDK
+	bin/$(tfgen) nodejs \
+		--overlays provider/overlays/nodejs \
+		--out sdk/nodejs
+	cd sdk/nodejs
+	yarn install
+	yarn run tsc
+	cp $(provider_readme) ./bin/README.md
+	cp ../../LICENSE \
+		package.json \
+		yarn.lock \
+		./bin/
+	sed -i.bak \
+		-e "s/\$${VERSION}/$(nodejs_version)/g" \
+		./bin/package.json
```

Comment:
why this -i.bak if you just immediately rm it

Reply:
This came in with the boilerplate that Pulumi provides... I _think_ it's because they work on Macs; I seem to recall some difference in behavior with the `sed` that comes with macOS that requires this.

Basically, a cross-platform support kinda thing. I can remove it or keep it.

Reply:
Got it. It's fine, just seemed weird on initial read.

Reply:
Yeah, pretty much all these SDK targets are exactly what Pulumi's generator made (with the exception of the `npm pack` bit on the NodeJS one, which I added... easier to deal with tarballs for both Python and NodeJS, rather than a tarball for one and a directory of loose files for the other.)